### PR TITLE
Update installed caskfiles of modified casks.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Often casks get updated without a version bump to fix a broken `uninstall` or add something to `zap`. This replaces the installed caskfile if it differs from the current one and the version hasn't changed.

### CODE REVIEW

File: Library/Homebrew/cmd/update-report.rb [hunk 1]
```diff
@@ -572,6 +575,24 @@ def migrate_tap_migration
...
+      current_version = cask.version
+      installed_version = cask.versions.last
+
+      current_caskfile = cask.sourcefile_path
+      installed_caskfile = cask.installed_caskfile
+
+      next unless current_version == installed_version
+      next if current_caskfile.sha256 == installed_caskfile.sha256
+
+      installed_caskfile.atomic_write File.read(current_caskfile)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: Should handle if the write (or perhaps even read) fails here.

**Comment by MikeMcQuaid** [comment]
What's the overhead on doing this on every `brew update`?

**Comment by vitorgalvao** [comment]
Big thumbs up to this. Not only will it fix issues with uninstalls, it should help prevent problems like the `undent` deprecation (closes https://github.com/Homebrew/homebrew-cask/issues/49716).

**Comment by MikeMcQuaid** [comment]
> it should help prevent problems like the undent deprecation (closes Homebrew/homebrew-cask#49716).

It won't prevent that because if the `version` of a Cask is updated it'll still have the same problem.

**Comment by vitorgalvao** [comment]
> It won't prevent that because if the `version` of a Cask is updated it'll still have the same problem.

On the HBC repos, it’s customary that when we introduce a change like the `undent` deprecation that affects a lot of casks, we make a PR that introduces the change (and nothing else) to all relevant casks at once.

**Comment by MikeMcQuaid** [comment]
> On the HBC repos, it’s customary that when we introduce a change like the undent deprecation that affects a lot of casks, we make a PR that introduces the change (and nothing else) to all relevant casks at once.

Even when you do that: it's still a problem. If you're relying on having ambiguously old Cask files which use Homebrew APIs on them: eventually some of those APIs will be deprecated, disabled and removed and you'll have the same problem again. Assuming that the current version of Cask has infinite backwards compatibility is the root issue here.

**Comment by reitermarkus** [comment]
Yes, I am aware that this won't fix the problem. There really isn't a fix for this, other than never deprecating anything. However, this will prevent deprecation errors for people who run `brew update` regularly.

Even I had one cask still using `.undent`, and I run `brew update` daily.

**Comment by MikeMcQuaid** [comment]
> There really isn't a fix for this, other than never deprecating anything.

I disagree. Several options off the top of my head:

- Catch these exceptions and fall back to the current Cask file's `uninstall`
- Catch these exceptions and print a warning saying uninstall hooks are not run
- Instead of storing the entire Cask file in the Caskroom store just the part that's needed e.g. the uninstall hook method
- Instead of storing the entire Cask file in the Caskroom store instead store something to define uninstall logic that doesn't rely on Homebrew internals e.g. a Bash script or a JSON file that defines logic

**Comment by commitay** [comment]
> Catch these exceptions and fall back to the current Cask file's uninstall

https://github.com/Homebrew/homebrew-cask/issues/49716#issuecomment-406605565

**Comment by MikeMcQuaid** [comment]
@commitay I replied to that comment there. Can you be more explicit on what you're trying to say rather than just dropping a single link to a comment you'd made previously? Would you care to address any of my other options?

To be more explicit on my viewpoint: no-one except me seems to be suggesting solutions to this problem which will not recur next time we remove formerly deprecated then disabled Homebrew APIs. Homebrew formulae have solutions to this problem. I'm not pretending I have all the answers but the current situation of "it'll happen again and there's nothing we can do about it" is defeatist and makes it hard for us to remove any APIs in future.

**Comment by vitorgalvao** [comment]
> Catch these exceptions and fall back to the current Cask file's `uninstall`

A bunch of casks have changes that make this impractical. E.g., software that was a `.app` is now a `.pkg`. Or something that didn’t have an uninstall script, now has one. I don’t have numbers on how often this happens, but it’s not incredibly rare.

> Catch these exceptions and print a warning saying uninstall hooks are not run

Worse experience for the user, but I’m not opposed to it. We’d need to document why this happens and explain how to uninstall manually, but it might work.

> Instead of storing the entire Cask file in the Caskroom store just the part that's needed e.g. the uninstall hook method

Impractical if the deprecation is on a part needed for the uninstall.

> Instead of storing the entire Cask file in the Caskroom store instead store something to define uninstall logic that doesn't rely on Homebrew internals e.g. a Bash script or a JSON file that defines logic

I like this idea.

**Comment by MikeMcQuaid** [comment]
> Worse experience for the user, but I’m not opposed to it. We’d need to document why this happens and explain how to uninstall manually, but it might work.

I think it's a better experience than the current situation of an unhandled but expected error.

> Impractical if the deprecation is on a part needed for the uninstall.

Yep but it still reduces the "surface" somewhat.



> I like this idea.

🎉

**Comment by reitermarkus** [comment]
> - Catch these exceptions and fall back to the current Cask file's `uninstall`
>  - Catch these exceptions and print a warning saying uninstall hooks are not run

Those are not fixes, those are fallbacks.

> - Instead of storing the entire Cask file in the Caskroom store just the part that's needed e.g. the uninstall hook method
> - Instead of storing the entire Cask file in the Caskroom store instead store something to define uninstall logic that doesn't rely on Homebrew internals e.g. a Bash script or a JSON file that defines logic

This sounds good in theory, but I'm not sure if this isn't just shifting the problem elsewhere and additionally making the uninstall less flexible.

**Comment by MikeMcQuaid** [comment]
> Those are not fixes, those are fallbacks.

Indeed but they are an improvement on the current situation of a known, unhandled exception.

> This sounds good in theory, but I'm not sure if this isn't just shifting the problem elsewhere and additionally making the uninstall less flexible.

It will make uninstall less flexible but it's not going to increase the surface area of problems. In the `undent` case: most of these issues seem to be coming from places other than the uninstall block. Additionally, a bash script would isolate entirely from Homebrew Cask's Ruby (or even Bash) logic as Apple aren't going to be breaking Bash's backwards compatibility any time soon.

**Comment by MikeMcQuaid** [comment]
@Homebrew/cask any thoughts or updates on this? Please note as above these DSLs will be removed in Homebrew 2.0 around February regardless of a PR like this being merged.

**Comment by vitorgalvao** [comment]
I’m still fine with the idea behind this.

**Pull request closed without merging**